views.go

```diff
@@ -330,27 +330,79 @@ func (lk *logKeeper) appendGlobalLog(w http.ResponseWriter, r *http.Request) {
 	lk.render.WriteJSON(w, http.StatusCreated, createdResponse{"", testUrl})
 }
 
-func (lk *logKeeper) viewBuildById(w http.ResponseWriter, r *http.Request) {
-	w.Header().Add("Access-Control-Allow-Origin", "*")
-	defer r.Body.Close()
+func (lk *logKeeper) viewBuildByIdInS3(r *http.Request, buildID string) (*model.Build, []model.Test, *apiError) {
+	var wg sync.WaitGroup
+	wg.Add(2)
 
-	vars := mux.Vars(r)
-	buildID := vars["build_id"]
+	var build *model.Build
+	var buildErr error
+	go func() {
+		defer recovery.LogStackTraceAndContinue("fetching build from s3 for test id")
+		defer wg.Done()
+		build, buildErr = lk.opts.Bucket.FindBuildByID(r.Context(), buildID)
+	}()
 
+	var tests []model.Test
+	var testsErr error
+	go func() {
+		defer recovery.LogStackTraceAndContinue("fetching build from s3 for test id")
+		defer wg.Done()
+		tests, testsErr = lk.opts.Bucket.FindTestsForBuild(r.Context(), buildID)
+	}()
+
+	wg.Wait()
+
+	if buildErr != nil {
+		lk.logErrorf(r, "Error finding build: %v", buildErr)
+		return nil, nil, &apiError{Err: "failed to find build in S3:" + buildErr.Error(), code: http.StatusInternalServerError}
+	}
+	if build == nil {
+		return nil, nil, &apiError{Err: "view build: build not found in S3", code: http.StatusNotFound}
+	}
+	if testsErr != nil {
+		lk.logErrorf(r, "Error finding tests for build: %v", testsErr)
+		return nil, nil, &apiError{Err: testsErr.Error(), code: http.StatusInternalServerError}
+	}
+	return build, tests, nil
+}
+
+func (lk *logKeeper) viewBuildByIdInDatabase(r *http.Request, buildID string) (*model.Build, []model.Test, *apiError) {
 	build, err := model.FindBuildById(buildID)
 	if err != nil {
 		lk.logErrorf(r, "Error finding build: %v", err)
-		lk.render.WriteJSON(w, http.StatusInternalServerError, apiError{Err: "failed to find build:" + err.Error()})
-		return
+		return nil, nil, &apiError{Err: "failed to find build:" + err.Error(), code: http.StatusInternalServerError}
```

Comment:
I think you need a space: `"failed to find build: "+ err.Error()`.

Reply:
Switching to using string formatting here anyway.